src/server/pps/server/api_server.go

```diff
@@ -4139,3 +4139,64 @@ func (a *apiServer) SetProjectDefaults(ctx context.Context, req *pps.SetProjectD
 	}
 	return &resp, nil
 }
+
+// PipelinesSummary implements the protobuf pps.PipelinesSummary RPC
+func (a *apiServer) PipelinesSummary(ctx context.Context, req *pps.PipelinesSummaryRequest) (*pps.PipelinesSummaryResponse, error) {
+	var projects []*pfs.Project
+	for _, p := range req.Projects {
+		var project *pfs.Project
+		switch p.Picker.(type) {
+		case *pfs.ProjectPicker_Name:
+			project = &pfs.Project{Name: p.GetName()}
+		default:
+			return nil, errors.Errorf("project picker is of an unknown type: %T", p.Picker)
+		}
+		if err := a.env.AuthServer.CheckProjectIsAuthorized(ctx, project, auth.Permission_PROJECT_LIST_REPO); err != nil {
```

Comment:
Do we care about a time-of-check/time-of-use issue here? We could check permissions here, then the user’s permissions on a project could be revoked, then pipeline state could change, then `ListPipelineInfo` could run, then we would return summary information about the revoked project. Of note, we potentially leak pipeline names below at https://github.com/pachyderm/pachyderm/pull/9889/files#diff-e400d3c661e4039b9f23301b99e8a22707cf3fdfa18865f569b499d3d708840dR4185 (imagine if the pipeline created immediately after revoking `alice`’s access is `fire-alice` …).

Given that `ListPipelineInfo` processes all pipelines anyway, maybe the check should be used when collecting results? Or even done twice, once to validate the request and once to sanitize the response?

Or maybe we don’t care, in which case it’s probably a good idea to document it.

Reply:
I recognize the concern but given that we're already doing this in many places, and the risk of error is extremely unlikely, we should document this and revisit if we overhaul the approach in the future. It may make sense to revisit this once PPS is migrated off of collections and many of these operations could be executed in a single query